django/thunderstore/usermedia/api/experimental/views/upload.py

```diff
@@ -0,0 +1,178 @@
+from datetime import timedelta
+
+from botocore.exceptions import ClientError
+from django.utils import timezone
+from drf_yasg.utils import swagger_auto_schema
+from rest_framework import status
+from rest_framework.generics import GenericAPIView
+from rest_framework.permissions import IsAuthenticated
+from rest_framework.response import Response
+
+from thunderstore.usermedia.api.experimental.serializers import (
+    UserMediaCreatePartUploadUrlsParams,
+    UserMediaFinishUploadParamsSerializer,
+    UserMediaInitiateUploadParams,
+    UserMediaSerializer,
+    UserMediaUploadUrlsSerializer,
+)
+from thunderstore.usermedia.exceptions import InvalidUploadStateException
+from thunderstore.usermedia.models import UserMedia
+from thunderstore.usermedia.s3_client import get_s3_client
+from thunderstore.usermedia.s3_upload import (
+    abort_upload,
+    create_upload,
+    finalize_upload,
+    get_signed_upload_urls,
+)
+
+PART_SIZE = 1024 * 1024 * 50
+
+
+class UserMediaInitiateUploadApiView(GenericAPIView):
+    queryset = UserMedia.objects.active()
+    serializer_class = UserMediaSerializer
+    # TODO: Add test for permission check
+    permission_classes = [IsAuthenticated]
+
+    @swagger_auto_schema(
+        request_body=UserMediaInitiateUploadParams,
+        responses={201: UserMediaSerializer()},
+        operation_id="experimental.usermedia.initiate-upload",
+    )
+    def post(self, request, *args, **kwargs):
+        validator = UserMediaInitiateUploadParams(data=request.data)
+        validator.is_valid(raise_exception=True)
+        total_size = validator.validated_data["file_size_bytes"]
+
+        client = get_s3_client()
+        user_media = create_upload(
+            client=client,
+            user=request.user,
+            filename=validator.validated_data["filename"],
+            size=total_size,
+            expiry=timezone.now() + timedelta(days=1),
+        )
+        serializer = self.get_serializer(user_media)
+        return Response(
+            serializer.data,
+            status=status.HTTP_201_CREATED,
+        )
+
+
+class UserMediaCreatePartUploadUrlsApiView(GenericAPIView):
+    queryset = UserMedia.objects.active()
+    lookup_field = "uuid"
+    lookup_url_kwarg = "uuid"
+    serializer_class = UserMediaCreatePartUploadUrlsParams
+    # TODO: Add test for permission check
+    permission_classes = [IsAuthenticated]
+
+    @swagger_auto_schema(
+        request_body=UserMediaCreatePartUploadUrlsParams,
+        responses={200: UserMediaUploadUrlsSerializer()},
+        operation_id="experimental.usermedia.create-part-upload-urls",
+    )
+    def post(self, request, *args, **kwargs):
+        instance = self.get_object()
+
+        validator = self.get_serializer(data=request.data)
+        validator.is_valid(raise_exception=True)
+
+        total_size = validator.validated_data["file_size_bytes"]
+        part_count = -(-total_size // PART_SIZE)
```

Comment:
This would round the result to floor as opposed to ceil, which is what the double negative is for

Reply:
This should probably be commented/made into a util/replaced with `math.ceil` (which has issues with very large numbers but we shouldn't get close to that) since it's not very intuitive.

Reply:
Then you'd have to convert it back to an integer anyway, so might as well use integer math to begin with. I'd rather add a comment for clarity than rely on math.ceil here